### PR TITLE
[SPARK-42495][CONNECT] Scala Client add Misc, String, and Date/Time functions

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds the following functions to the scala client:
- Misc functions.
- String functions.
- Date/Time functions.

### Why are the changes needed?
We want to be able the same APIs in the scala client as in the original Dataset API.

### Does this PR introduce _any_ user-facing change?
Yes, it adds new for functions to the Spark Connect Scala client.

### How was this patch tested?
Added tests to `PlanGenerationTestSuite` (and indirectly to `ProtoToPlanTestSuite`). Overloads are tested in `FunctionTestSuite`.

### CODE REVIEW

**Review by zhengruifeng — APPROVED** [review]

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -2238,6 +2238,1359 @@ object functions {
...
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  // Misc functions
+  //////////////////////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Calculates the MD5 digest of a binary column and returns the value as a 32 character hex
+   * string.
+   *
+   * @group misc_funcs
+   * @since 1.5.0
```
Comment (dongjoon-hyun) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:2250]: Ur, I'm a little confused. Is it correct, @hvanhovell and @zhengruifeng ?

cc @HyukjinKwon

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
This PR seems to be inconsistent from our code base. Could you explain what is the policy of this file? The following is the result from `master` branch.
```
$ grep since connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala | uniq -c
   1  * @since 3.4.0
 221    * @since 3.4.0
```

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:2250]: yeah let's change `since`

**Comment by hvanhovell** [comment]
I forgot to update it. TBH since these are verbatim copies of the existing API, I think there is a case to be made to use the original version instead the version of when the client was built.

**Comment by HyukjinKwon** [comment]
Or we can remove all `@since *` and just add one `@since 3.4.0` at the top level of the connect. And then, let the new APIs to have `@since 3.5.0` in the next release.

**Comment by HyukjinKwon** [comment]
Or just grip and replace `@since *` to `@since 3.4.0` ...

**Comment by dongjoon-hyun** [comment]
When we add new languages like Python/R, did you use the same versions of Scala API, @hvanhovell ? They were also a wrappers on top of Scala API.

> I forgot to update it. TBH since these are verbatim copies of the existing API, I think there is a case to be made to use the original version instead the version of when the client was built.

**Comment by LuciferYang** [comment]
Personally, I think it is a little strange to change `@since `. For example,  the following function is changed to `@since 3.4.0`, but `deprecated` since 2.1.0 ....

https://github.com/apache/spark/blob/1688a8768fb34060548f8790e77f645027f65db2/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala#L221-L226

**Comment by dongjoon-hyun** [comment]
That's a good question. Maybe, what about avoiding this propagation of deprecated methods, @hvanhovell , @HyukjinKwon , @LuciferYang ? Since we already deprecated these, it sounds more consistent with what Apache Spark has been claiming (not to use this APIs in new use cases).

**Comment by LuciferYang** [comment]
+1, Agree `avoiding this propagation of deprecated methods`

**Comment by hvanhovell** [comment]
I have updated the since tags.

A couple of things I would like to highlight. The scala clients tries to be as compatible as possible (both source and binary) with the exiting API. That is why this is different that API that is built on top of the scala API, that was new, this however is not, it should be or at least aspires to be the same API. Hence my argument that you could also use the original since tags. Anyway I am fine with updating them, like I did in other PRs, I just missed it in this case.

As for the removal of deprecated methods. Again we want to be as compatible as possible, so I want to given users access to them. We aspire that in the best case a user can migrate their code to connect without any issues (preferably no recompilation). I don't think we should ever remove these methods, they are poorly named that is all. They do not inhibit us in any way to keep evolving Spark since they are just aliases. I have kept out a few deprecated methods in the current connect Dataset, but I do intent to bring them back once we can support them.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for update, @hvanhovell .
Thank you all!

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala [hunk 2]
```diff
@@ -115,7 +116,7 @@ class ProtoToParsedPlanTestSuite extends SparkFunSuite with SharedSparkSession {
       val planner = new SparkConnectPlanner(spark)
       val catalystPlan =
         analyzer.executeAndCheck(planner.transformRelation(relation), new QueryPlanningTracker)
-      val actual = normalizeExprIds(catalystPlan).treeString
+      val actual = normalizeExprIds(ReplaceExpressions(catalystPlan)).treeString
```
Comment (cloud-fan) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala:119]: @hvanhovell is there a specific reason to replace `RuntimeReplaceable` in the golden files? It makes the golden file a bit hard to read as there are quite some expressions that will be replaced with `StaticInvoke`/`Invoke` and you will see something like
```
Project [staticinvoke(class org.apache.spark.sql.catalyst.expressions.ExpressionImplUtils, BinaryType, aesEncrypt, cast(g#0 as binary), cast(g#0 as binary), g#0, DEFAULT, cast( as binary), cast( as binary), BinaryType, BinaryType, StringTypeAnyCollation, StringTypeAnyCollation, BinaryType, BinaryType, true, true, true) AS aes_encrypt(g, g, g, DEFAULT, , )#0]
```

Comment (cloud-fan) [reply connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala:119]: Also for testing, I think it's good enough to verify the analyzed plan of Spark Connect.

Also cc @HyukjinKwon

**Pull request closed without merging**